From 0e7182f6e136ef55decc42552ca2a71f9b55db2f Mon Sep 17 00:00:00 2001
From: Alan Wang <wbo@codeaurora.org>
Date: Fri, 6 May 2016 17:52:57 +0800
Subject: [PATCH 23/31] Enable io-mode for DMABUF: correct the codes in
 v4l2bufferpool that don't use buffer-copying when we dealing with
 GST_V4L2_IO_MMAP. And make sure that this is the unique scenario.

Change-Id: I0d5d9735f8aa0173971bb724ee5a6e56396f8541
Signed-off-by: Alan Wang <wbo@codeaurora.org>
---
 sys/v4l2/gstv4l2bufferpool.c | 11 +++++++----
 1 file changed, 7 insertions(+), 4 deletions(-)

diff --git a/sys/v4l2/gstv4l2bufferpool.c b/sys/v4l2/gstv4l2bufferpool.c
index 0c68c79..1366c71 100644
--- a/sys/v4l2/gstv4l2bufferpool.c
+++ b/sys/v4l2/gstv4l2bufferpool.c
@@ -1877,7 +1877,8 @@ gst_v4l2_buffer_pool_process (GstV4l2BufferPool * pool, GstBuffer ** buf)
           GstV4l2MemoryGroup *group;
           gint index;
 
-          if ((*buf)->pool != bpool) {
+          if ((*buf)->pool != bpool &&
+                  ((obj->mode) != GST_V4L2_IO_MMAP)) {
             /* APBU-10143 Todo.
                Temp disable this check, the buffer from gst-omx will not
                include pool value. but it should not goto copying.
@@ -1885,7 +1886,7 @@ gst_v4l2_buffer_pool_process (GstV4l2BufferPool * pool, GstBuffer ** buf)
             GST_WARNING_OBJECT(pool,
               "pool in gstbuffer is not equal to v4l2 pool");
 
-            //goto copying;
+            goto copying;
           }
 
           if (!gst_v4l2_is_buffer_valid (*buf, &group))
@@ -1953,7 +1954,8 @@ gst_v4l2_buffer_pool_process (GstV4l2BufferPool * pool, GstBuffer ** buf)
           /* Remove our ref, we will still hold this buffer in acquire as needed,
            * otherwise the pool will think it is outstanding and will refuse to stop. */
           if (gst_mini_object_get_qdata (GST_MINI_OBJECT_CAST (to_queue),
-                GST_OMX_VIDEO_DEC_OMX_BUFFER_QUARK) == NULL) {
+                GST_OMX_VIDEO_DEC_OMX_BUFFER_QUARK) == NULL ||
+                  ((obj->mode) != GST_V4L2_IO_MMAP)) {
             gst_buffer_unref (to_queue);
           }
 
@@ -1966,7 +1968,8 @@ gst_v4l2_buffer_pool_process (GstV4l2BufferPool * pool, GstBuffer ** buf)
               /* release the rendered buffer back into the pool. This wakes up any
                * thread waiting for a buffer in _acquire(). */
               if (gst_mini_object_get_qdata (GST_MINI_OBJECT_CAST (out),
-                    GST_OMX_VIDEO_DEC_OMX_BUFFER_QUARK) == NULL)
+                    GST_OMX_VIDEO_DEC_OMX_BUFFER_QUARK) == NULL ||
+                      ((obj->mode) != GST_V4L2_IO_MMAP))
                 gst_v4l2_buffer_pool_release_buffer (bpool, out);
               else
                 /* This buffer is copied from the allocated buffer, so we can not
-- 
1.9.1

